src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -1098,28 +1253,56 @@ bool is_object_properties_superset(
             auto regex = re2::RE2(as_string_view(propPattern));
             if (re2::RE2::PartialMatch(pname, regex)) {
                 pattern_match_found = true;
-                if (!is_superset(ctx, schemaPattern, schema)) {
-                    // not compatible
-                    return false;
+
+                auto prop_res = is_superset(
+                  ctx, schemaPattern, schema, prop_path);
+
+                if (prop_res.has_error()) {
+                    res.merge(std::move(prop_res));
+                    res.emplace<json_incompatibility>(
+                      std::move(prop_path),
+                      json_incompatibility_type::
+                        property_removed_not_covered_by_partially_open_content_model);
                 }
+                break;
             }
         }
+        if (pattern_match_found) {
+            // check next property
+            continue;
+        }
 
         // or it should check against older["additionalProperties"], if no match
         // in patternProperties was found
-        if (
-          !pattern_match_found
-          && !is_superset(ctx, older_additional_properties, schema)) {
-            // not compatible
-            return false;
+        if (!is_false_schema(older_additional_properties)) {
+            auto add_prop_res = is_superset(
+              ctx, older_additional_properties, schema, prop_path);
+
+            if (add_prop_res.has_error()) {
+                res.merge(std::move(add_prop_res));
+                res.emplace<json_incompatibility>(
+                  std::move(prop_path),
+                  json_incompatibility_type::
+                    property_removed_not_covered_by_partially_open_content_model);
+            }
+            continue;
```

Comment:
I don't follow the scope here, either.

Reply:
I think this one's correct, let me know if you disagree. The check above for `!is_false_schema(older_additional_properties)` tells us whether additional properties are allowed. If they are (ie. this if branch) then regardless of whether `is_superset` gives an error or not, we're done with evaluating this property.

Would a refactoring like this would make things less/more confusing? https://github.com/redpanda-data/redpanda/commit/a7ff3115d5c7508ea9f0ab76eb6af4227182346a

Reply:
Oh, I see.  No, this is fine.